proposals/extend_cli_for_generate_policies.md

```diff
@@ -0,0 +1,130 @@
+
+#### Example: Add Network Policy
+
+A sample test file for [this policy](https://kyverno.io/policies/best-practices/add_network_policy/) will look like this -
+
+```yaml
+name: deny-all-traffic
+policies:
+  - policy.yaml
+resources:
+  - resource.yaml
+results:
+  - policy: add-networkpolicy
+    rule: default-deny
+    resource: hello-world-namespace
+    generatedResource: generatedResource.yaml
+    kind: Namespace
+    result: pass
+```
+resource.yaml is the manifest for creating a new Namespace resource. The result will pass if the generatedResource.yaml looks like this -
+
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: default-deny
+  namespace: hello-world-namespace
+spec:
+  podSelector: {}
+  policyTypes:
+  - Ingress
+  - Egress
+```
+
+### Method to test clone object
+
+Generate rules with clone object requires a pre-existing resource in the cluster. The manifest of the rule will not have the definition of this resource. Instead, the rule will only specify the name of the resource and the namespace where it resides. To have the resource exist on the cluster before the rule is applied, user should provide the definition of the resource. The rest of the approach remains same as that of data object.
+
+#### Example: Sync Secrets
+
+Consider [this policy](https://kyverno.io/policies/other/sync_secrets/). The test file should look like this -
+
+```yaml
+name: sync-secrets
+policies:
+  - policy.yaml
+resources:
+  - resource.yaml
+results:
+  - policy: sync-secrets
+    rule: sync-image-pull-secret
+    resource: hello-world-namespace
+    generatedResource: generatedResource.yaml
+    cloneResource: cloneResource.yaml # Pre-existing resource definition
+    kind: Namespace
+    result: pass
+```
+
+User should provide the pre-existing resource in the cloneResource field of the results array. The cloneResource.yaml for the above test file will look like this -
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: regcred
+  namespace: default
+type: Opaque
+data:
+  password: MWYyZDFlMmU2N2Rm
+```
+
+The test will pass if the generatedResource.yaml looks like this -
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: regcred
+  namespace: hello-world-namespace
+type: Opaque
+data:
+  password: MWYyZDFlMmU2N2Rm
+```
+
+## Implementation details
+
+Most of the implementation is similar to testing mutate policies. The key difference is that generate rules require Generate-Request Controller.
+
+### Mock Generate-Request Controller
+
+To process generate rules, we need a Generate-Request Controller. We can make a simple instance of the Controller with only the Kubernetes' dynamic client. The dynamic client must be a fake one so as to not process the rule on an actual cluster. Additionally, the triggering resource and pre-existing resource for the clone object should exist in the cluster. The dynamic client will be initialized with these resources.
```

Comment:
Based on the above proposal, all test resources are local resources, do we need the dynamic client?